### PR TITLE
fix(hook): trace Workflow-spawned subagent transcripts

### LINKED ISSUE #43: Workflow-spawned subagents are invisible: metas live under subagents/workflows/<run>/ and carry no toolUseId
## Summary

Subagents spawned by the `Workflow` **tool** produce no observations at all — no span, no generation, no token usage. A single `Workflow` call fans out to dozens of agents, all running `sonnet`, and none of that work reaches Langfuse. For token-metering and cost use cases this is the same silent undercount as langfuse/claude-observability-plugin#28, but through a different code path and with an extra structural wrinkle.

Measured on one workflow run: **40 agent transcripts, 40 metas, zero observations emitted.** In the 100 most recent observations of a project that runs workflows heavily, every `GENERATION` is `claude-opus-5` — not a single `sonnet` generation exists, which is the symptom to look for.

## Root cause

**1. Discovery is not recursive, and workflow metas are two levels down.**

`get_subagent_transcripts_by_tool_use_id` globs a single level:

```python
subagent_dir = transcript_path.with_suffix("") / "subagents"
for meta_path in subagent_dir.glob("*.meta.json"):
```

In a session that uses workflows, `subagents/` contains only a `workflows` directory; the metas live at:

```
subagents/workflows/wf_<run-id>/agent-<agent-id>.meta.json
```

`glob("*.meta.json")` never descends, so there are zero matches.

**2. The meta carries no** `toolUseId` — same skip as langfuse/claude-observability-plugin#28, different meta shape:

```json
{"agentType":"workflow-subagent","spawnDepth":1,"model":"sonnet"}
```

No `toolUseId`, no `description`, no `name`. The `continue` drops it.

Both defects must be fixed for either to matter: a recursive glob alone still hits the `toolUseId` skip, and relaxing the skip alone never reaches the files.

**Note for** [#29](<https://github.com/langfuse/claude-observability-plugin/issues/29>)**:** `discover_teammate_transcripts` in that PR uses the same single-level `subagent_dir.glob("*.meta.json")`, so as it stands it would not pick these up either. A recursive walk (or an explicit `subagents/workflows/*/` pass) would let one discovery path serve both cases.

## The structural wrinkle: N agents per `tool_use_id`

Unlike classic subagents, workflow agents are **many-to-one**: the 40 agents above all belong to a single `Workflow` tool call. The existing map is one-to-one —

```python
subagent_transcripts_by_tool_use_id[tool_use_id] = {"path": jsonl_path, ...}
```

— so they cannot simply be keyed into it; they would overwrite each other. A synthetic key does not help either, because attachment happens while emitting the observation for a **real** `tool_use_id` found in the transcript.

The parallel-path approach [#29](<https://github.com/langfuse/claude-observability-plugin/issues/29>) takes for teammates (separate discovery + separate emission + its own progress entry in the same state store) looks like the right shape here too, and avoids restructuring the 1:1 map.

## The linkage does exist

Everything needed to attach a workflow run to its launching tool call is already in the parent transcript. The `<task-notification>` row carries the tool use id and the run id in the same block:

```
<task-id>wpusx6utx</task-id>
<tool-use-id>toolu_01F21Lx6jhFshwLHFLpeHNuC</tool-use-id>
… wf_86dadd3f-a52 …
```

so `wf_<run-id> → toolu_…` is derivable, and the hook already parses these rows for async agent deferral (`is_task_notification_row`, `pending_task_notifications`).

Per-run there is also a `journal.jsonl` next to the agent transcripts, with one `started` and one `result` event per agent:

```json
{"type":"started","key":"v2:<sha256>","agentId":"a34f114e34fc6f6ec"}
{"type":"result","key":"v2:<sha256>","agentId":"a34f114e34fc6f6ec","result":"…"}
```

`agentId` matches the `agent-<agent-id>.jsonl` filenames, so completion is detectable per agent, and the run is complete when every `started` has its `result` — useful for the deferral decision, since the hook fires on `Stop` while a background workflow may still be running.

## Open question: what to name them

There is no per-agent label anywhere in the persisted data:

* `agentType` is always the literal `workflow-subagent`
* the journal's `key` is a `v2:<sha256>` content hash, not a label
* the agent transcript's first row has `entrypoint: "cli"`, `isSidechain: true`, `parentUuid: null` — nothing human-readable

The workflow script does define labels (`agent(prompt, {label, phase})`), and the generated script is saved alongside the run at `workflows/scripts/<name>-wf_<run-id>.js`, but the per-agent label is not persisted into the meta or the journal.

That leaves a real trade-off, and it may need a change in Claude Code rather than in the plugin:

* naming every agent `workflow-subagent` collapses an entire fan-out into one Agent Graph node
* naming by prompt is high-cardinality, which the [best-practices guide](<https://langfuse.com/docs/observability/best-practices>) explicitly warns against

Even with an imperfect name, emitting these observations at all would fix the cost/token blind spot, which is the more damaging half of the bug.

## Environment

Plugin 1.0.0 (marketplace install), `requires-python >=3.10`, macOS. Reproduced with `Workflow`-based fan-out under a `Stop`-triggered hook run.

Related: langfuse/claude-observability-plugin#28 / [#29](<https://github.com/langfuse/claude-observability-plugin/issues/29>) (same skip, teammate metas), langfuse/claude-observability-plugin#5 (`toolUseResult.agentId` fallback).

---

*Provenance: the figures above are measured on a live installation (plugin 1.0.0, macOS), not estimated. The agent and meta counts come from the on-disk transcripts of a real workflow run; the observation and model breakdown comes from the Langfuse API for the same project over the same window. Happy to provide the exact commands, or a redacted* `meta.json` */* `journal.jsonl` *sample, if that helps reproduce it.*

### PR DESCRIPTION
**Pull request by milanagm**
## Summary

Agents spawned by the `Workflow` tool write transcripts to `<session>/subagents/workflows/wf_<runId>/agent-*.jsonl`, with metas of exactly `{"agentType":"workflow-subagent","spawnDepth":1}`. The hook never saw them, for three independent reasons:

1. discovery globs `subagents/*.meta.json` non-recursively — one level too high
2. metas without `toolUseId` are skipped — workflow metas never have one
3. the async-hold whitelist only knows `Agent`/`Task`, so a running workflow never holds the turn open (the root span froze before the workflow finished, and the `Tool: Workflow` span stayed a seconds-long stub)

## Fix

* new discovery pass for the workflow layout, keyed by run directory (`wf_<runId>`), accepting metas without `toolUseId`; classic discovery untouched
* launch linkage via the structured `toolUseResult` (`taskType == "local_workflow"`, `runId`)
* `"Workflow"` added to the async-hold whitelist; the completion notification resolves the turn through the existing `<tool-use-id>` machinery
* each agent emits nested under its `Tool: Workflow` span with stable cursor keys (`workflow-agent:<tool_use_id>:<agent_id>`) — idempotent across Stop firings; journal results attach as span metadata; agent end times extend the tool span end
* two deliberate edge-case behaviors, both pinned by tests: agents whose final message carries no text fall back to their journal result as span output; interrupted agents (no assistant rows) emit no span

## Tests

140 pass (123 existing + 17 new): discovery unit tests (layout, missing/extra meta keys incl. the `worktreePath`/`model` variants, journal edge cases), deferral tests (running workflow holds the turn, notification resolves it, launch text alone never links), and end-to-end emission tests (nesting asserted via span parents, token arithmetic, cursor idempotence, a second run dir on disk stays out of the launching turn, truncated `<result>` payloads). The new fixture is harvested from real 2.1.215/2.1.221 transcripts and sanitized, structure byte-faithful.

Fixes langfuse/claude-observability-plugin#43 (langfuse/claude-observability-plugin#43).

### CODE REVIEW

**Comment by hassiebp** [comment]
@claude review

File: hooks/langfuse_hook.py [hunk 1]
```diff
@@ -1370,6 +1401,78 @@ def get_subagent_transcripts_by_tool_use_id(transcript_path: Path) -> Dict[str,
...
+                continue
+            if not isinstance(metadata, dict) or metadata.get("agentType") != "workflow-subagent":
+                continue
+
+            jsonl_path = meta_path.with_name(meta_path.name[: -len(".meta.json")] + ".jsonl")
+            if not jsonl_path.exists():
+                continue
+
+            agent_id = meta_path.name[: -len(".meta.json")]
+            if agent_id.startswith("agent-"):
```
[the remark this answers is not in the document]
Comment (milanagm) [reply hooks/langfuse_hook.py]: okay, makes sense. I created a helper method that contains the code lines, so that there are no duplications anymore.

File: hooks/langfuse_hook.py [hunk 2]
```diff
@@ -2629,6 +2859,13 @@ def emit_new_turns_from_transcript(
         if subagent_transcripts_by_tool_use_id:
             debug(f"Discovered {len(subagent_transcripts_by_tool_use_id)} subagent transcript(s)")
 
+        workflow_agent_transcripts_by_run_id = get_workflow_agent_transcripts_by_run_id(transcript_path)
```
[the remark this answers is not in the document]
Comment (milanagm) [reply hooks/langfuse_hook.py:2869]: I understand what you are saying. Each journal `result` row holds the full output of one agent. The hook reads the full journal on every Stop. So the repeated cost grows with the total workflow output. 

I suggest to still keep this PR @hassiebp as it is, for two reasons:
1. The classic discovery uses the same rescan pattern. This PR follows that existing design. 
2. A fix changes the data flow between discovery and emission. That change needs its own small PR.

The follow-up fix will do two things. It will read only the runs that the current turns reference. And it will read the journal only at emission time. I also want to check if we can skip those that were alredy fully emitted. 

Sessions without workflows pay no cost. The `workflows/` folder does not exist there, and the scan stops at once.

@hassiebp if you agree feel free to merge
